logic/transaction_pool.js

```diff
@@ -297,21 +297,24 @@ TransactionPool.prototype.getMergedTransactionList = function(reverse, limit) {
  *
  * @param {Object} transaction - Transaction object
  */
-TransactionPool.prototype.addUnconfirmedTransaction = function(transaction) {
-	if (
-		transaction.type === transactionTypes.MULTI ||
-		Array.isArray(transaction.signatures)
-	) {
-		self.removeMultisignatureTransaction(transaction.id);
-	} else {
-		self.removeQueuedTransaction(transaction.id);
-	}
+TransactionPool.prototype.addUnconfirmedTransaction = function(
+	transaction,
+	cb
+) {
+	__private.isMultisignatureTransaction(transaction, isMulti => {
+		if (isMulti) {
+			self.removeMultisignatureTransaction(transaction.id);
+		} else {
+			self.removeQueuedTransaction(transaction.id);
+		}
 
-	if (self.unconfirmed.index[transaction.id] === undefined) {
-		self.unconfirmed.transactions.push(transaction);
-		const index = self.unconfirmed.transactions.indexOf(transaction);
-		self.unconfirmed.index[transaction.id] = index;
-	}
+		if (self.unconfirmed.index[transaction.id] === undefined) {
+			self.unconfirmed.transactions.push(transaction);
+			const index = self.unconfirmed.transactions.indexOf(transaction);
+			self.unconfirmed.index[transaction.id] = index;
+		}
```

Comment:
This can be refactored later, we don't need to maintain two arrays for this purpose. Simply `self.unconfirmed.transactions` is good enough to maintain the unconfirmed transactions.